dev/benchmarks/c/ipc_benchmark.cc

```diff
@@ -0,0 +1,178 @@
+  benchmark::DoNotOptimize(schema);
+
+  nanoarrow::UniqueArrayView array_view;
+  NANOARROW_RETURN_NOT_OK(
+      ArrowArrayViewInitFromSchema(array_view.get(), schema.get(), nullptr));
+
+  while (true) {
+    nanoarrow::UniqueArray array;
+    NANOARROW_RETURN_NOT_OK(array_stream->get_next(array_stream, array.get()));
+    if (array->release == nullptr) {
+      break;
+    }
+
+    NANOARROW_RETURN_NOT_OK(
+        ArrowArrayViewSetArray(array_view.get(), array.get(), nullptr));
+
+    *batch_count = *batch_count + 1;
+  }
+
+  return NANOARROW_OK;
+}
+
+/// \defgroup nanoarrow-benchmark-ipc IPC Reader Benchmarks
+///
+/// Benchmarks for the ArrowArrayStream IPC reader.
+///
+/// @{
+
+/// \brief Use the ArrowArrayStream IPC reader to read 10,000 batches with 5 elements each
+/// from a file
+static void BenchmarkIpcReadManyBatchesFromFile(benchmark::State& state) {
+  int64_t batch_count = 0;
+  int64_t column_count = 0;
+
+  for (auto _ : state) {
+    nanoarrow::ipc::UniqueInputStream input_stream;
+    NANOARROW_THROW_NOT_OK(
+        MakeFixtureInputStreamFile("many_batches.arrows", input_stream.get()));
+
+    nanoarrow::UniqueArrayStream array_stream;
+    NANOARROW_THROW_NOT_OK(
+        ArrowIpcArrayStreamReaderInit(array_stream.get(), input_stream.get(), nullptr));
+
+    NANOARROW_THROW_NOT_OK(
+        ArrayStreamReadAll(array_stream.get(), &batch_count, &column_count));
+
+    benchmark::DoNotOptimize(batch_count);
+  }
+
+  state.SetItemsProcessed(state.items_processed() + batch_count);
+}
+
+/// \brief Use the ArrowArrayStream IPC reader to read 10,000 batches with 5 elements each
+/// from a buffer
+static void BenchmarkIpcReadManyBatchesFromBuffer(benchmark::State& state) {
+  int64_t batch_count = 0;
+  int64_t column_count = 0;
+
+  nanoarrow::UniqueBuffer buffer;
+  NANOARROW_THROW_NOT_OK(MakeFixtureBuffer("many_batches.arrows", buffer.get()));
+
+  for (auto _ : state) {
+    // Note: an attempt to remove this copy does not affect the timing for this particular
+    // benchmark (it is possible to set a deallocator that does nothing and manually
+    // assign the data and size_bytes of the copy).
+    nanoarrow::UniqueBuffer buffer_copy;
+    NANOARROW_THROW_NOT_OK(
+        ArrowBufferAppend(buffer_copy.get(), buffer->data, buffer->size_bytes));
+
+    nanoarrow::ipc::UniqueInputStream input_stream;
+    NANOARROW_THROW_NOT_OK(
+        ArrowIpcInputStreamInitBuffer(input_stream.get(), buffer_copy.get()));
+
+    nanoarrow::UniqueArrayStream array_stream;
+    NANOARROW_THROW_NOT_OK(
+        ArrowIpcArrayStreamReaderInit(array_stream.get(), input_stream.get(), nullptr));
+
+    NANOARROW_THROW_NOT_OK(
+        ArrayStreamReadAll(array_stream.get(), &batch_count, &column_count));
+
+    benchmark::DoNotOptimize(batch_count);
+  }
+
+  state.SetItemsProcessed(state.items_processed() + batch_count);
+}
+
+/// \brief Use the ArrowArrayStream IPC reader to read 10,000 columns with 0 batches from
```

Comment:
So this stresses reading the schema? Again, 0-length files or streams might exist, but they do not strike me as a valuable benchmark situation.

Reply:
Yes, you could also micro-benchmark the IPC decoder.